compiler/backend/src/org/jetbrains/kotlin/codegen/ExpressionCodegen.java

```diff
@@ -3107,6 +3108,18 @@ private StackValue generateEquals(
             );
         }
 
+        if ((opToken == KtTokens.EQEQ || opToken == KtTokens.EXCLEQ) &&
+            (isEnumClass(bindingContext.getType(left).getConstructor().getDeclarationDescriptor()) ||
```

Comment:
LHS should also be non-null. Use `TypeUtils.isNullableType` to check that type is (non-)nullable (it implements detailed check for possibly parameterized type.

The following test should pass:
```
enum class Test { A, B }

fun eq(x: Test?, y: Test) = x == y

fun box() = if (!eq(null, A)) "OK" else "Fail"
```